torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1546,6 +1608,28 @@ def _prep_grads_for_backward(self) -> None:
                 p.grad.size() != p._orig_size  # type: ignore[attr-defined]
                 or p.grad.device != p.device
             ):
+                offloaded: bool = p.grad.device != p.device
+                if offloaded:
+                    assert self.cpu_offload.offload_params, \
+                        "`p.grad.device` and `p.device` should be the same " \
+                        "if not offloading parameters to CPU"
+                prev_iter_outside_no_sync: bool = \
+                    p.grad.size() == p._local_shard.shape  # type: ignore[attr-defined]
+                # As long as the previous iteration was outside `no_sync()`,
+                # then we must save the gradient in `_saved_grad_shard`, even
+                # if the current iteration is inside `no_sync()`. This is to
+                # prepare for the next iteration outside `no_sync()`, which may
+                # try to accumulate gradients. FSDP accumulates gradients in
+                # the separate variable `p._saved_grad_shard` to leave `p.grad`
+                # for the per-iteration gradient.
+                if prev_iter_outside_no_sync:
+                    # FSDP currently does not support gradient accumulation
+                    # outside `no_sync()` when using CPU offloading. Trying to
+                    # do so yields incorrect results since FSDP will use the
+                    # newly-reduced gradient instead of accumulating with any
+                    # existing gradient.
+                    if not offloaded:
+                        p._saved_grad_shard = p.grad.data  # type: ignore[attr-defined]
```

Comment:
could we add a warning for this case and warning the comment "# FSDP currently does not support gradient accumulation
                    # outside `no_sync()` when using CPU offloading. Trying to
                    # do so yields incorrect results since FSDP will use the
                    # newly-reduced gradient instead of accumulating with any
                    # existing gradient."